orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy

```diff
@@ -26,43 +29,164 @@ import groovy.util.logging.Slf4j
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.stereotype.Component
 
+import java.util.concurrent.TimeUnit
+
+import static com.netflix.spinnaker.orca.ExecutionStatus.*
+
 @Component
 @Slf4j
-class ServerGroupCacheForceRefreshTask extends AbstractCloudProviderAwareTask implements Task {
-
+class ServerGroupCacheForceRefreshTask extends AbstractCloudProviderAwareTask implements RetryableTask {
   static final String REFRESH_TYPE = "ServerGroup"
 
+  long backoffPeriod = TimeUnit.SECONDS.toMillis(5)
+  long timeout = TimeUnit.MINUTES.toMillis(10)
+
   @Autowired
   OortService oort
 
+  @Autowired
+  ObjectMapper objectMapper
+
   @Override
   TaskResult execute(Stage stage) {
-    String cloudProvider = getCloudProvider(stage)
-    String account = getCredentials(stage)
-    Map<String, List<String>> capturedServerGroups = (Map<String, List<String>>) stage.context."deploy.server.groups"
+    def account = getCredentials(stage)
+    def cloudProvider = getCloudProvider(stage)
+    def stageData = stage.mapTo(StageData)
 
-    def zone = stage.context.zone
-    if (!zone) {
-      zone = stage.context.zones ? stage.context.zones[0] : null
+    def optionalTaskResult = performForceCacheRefresh(account, cloudProvider, stageData)
+    if (optionalTaskResult.present) {
+      return optionalTaskResult.get()
     }
 
-    def outputs = [:]
-    capturedServerGroups?.each { region, serverGroups ->
-      for (serverGroup in serverGroups) {
-        def model = [serverGroupName: serverGroup, asgName: serverGroup, region: region, account: account] // TODO retire asgName
+    boolean allAreComplete = processPendingForceCacheUpdates(account, cloudProvider, stageData, stage.startTime)
+    return new DefaultTaskResult(allAreComplete ? SUCCEEDED : RUNNING, objectMapper.convertValue(stageData, Map))
+  }
+
+  /**
+   * Every deployed server group should be force cache refreshed.
+   *
+   * An HTTP 200 response indicates that the force cache operation has completed and there is no need for additional
+   * polling. Long term, the expectation is that all caching agents will be asynchronous.
+   *
+   * An HTTP 202 response indicates that the force cache operation has been queued and will complete at some point in
+   * the near future.
+   */
+  private Optional<TaskResult> performForceCacheRefresh(String account, String cloudProvider, StageData stageData) {
+    def zone = stageData.zone
+
+    def refreshableServerGroups = stageData.deployServerGroups.collect { region, serverGroups ->
+      serverGroups.findResults { String serverGroup ->
+        def model = [asgName: serverGroup, serverGroupName: serverGroup, region: region, account: account]
```

Comment:
I've got an unreleased change in `clouddriver:ClusterCachingAgent` that'll let us get rid of `asgName` here.